### PR TITLE
ci: Type check perf

### PR DESCRIPTION
**Pull request by m-shaka**
### The author should do the following, if applicable
part of #3377 
- [ ] Add tests
- [ ] Run tests
- [ ] `bun run format:fix && bun run lint:fix` to format the code
- [ ] Add [TSDoc](https://tsdoc.org/)/[JSDoc](https://jsdoc.app/about-getting-started) to document the code

### detail
The main goal is to compare type check performances between main branch and each PR.

What CI jobs I added do is:
- generate Hono app with 200 routes
- compile a script using the app above and take diagnostics
- on main branch, store the result to GitHub actions cache
- on PR, compare the current performance with the latest stored result

I think it'd be nice to post the comparison as a comment on the PR (after making sure caching is working) 

### question
Should I do something for Codecov?

### CODE REVIEW

**Comment by yusukebe** [comment]
Hi @m-shaka !

This is great! It's unsure how this will be effective and how to measure performance accurately, but starting to try using it is important.

One thing about naming. How about creating `perf-measures` and putting this `type-check-perf` as `type-check`? In the future, we may create other scripts for measuring performance like `perf-measures/routers`.

**Comment by yusukebe** [comment]
> Should I do something for Codecov?

I'll investigate it, but it can be done after merging this.

**Comment by m-shaka** [comment]
Thank you for the feedback!
Could you explain more about what kind of files you are planning to place in `type-check` and `perf-measures` respectively?

**Comment by yusukebe** [comment]
> Could you explain more about what kind of files you are planning to place in `type-check` and `perf-measures` respectively?

We may not do that, and we don't have to do that, but I was thinking:

* Router speed
* Req => Res performance
* File/Bundle size like https://github.com/honojs/hono/pull/2975
* Other things in the [benchmarks](https://github.com/honojs/hono/tree/main/benchmarks)

**Comment by exoego** [comment]
We can ignore coverage for type perf code by adding
```
type-check-perf/**/*.ts
```
to
https://github.com/honojs/hono/blob/dfbd717263ab8ecd7bf495fbd2c32d8c79854284/vitest.config.ts#L19-L30

**Comment by yusukebe** [comment]
> We can ignore coverage for type perf code by adding

Yes, we should do it. Thank you for the advice.

**Comment by m-shaka** [comment]
Got it. You just mean `mv type-check-perf perf-measures/type-check`, don't you?

**Comment by yusukebe** [comment]
@m-shaka 

> Got it. You just mean `mv type-check-perf perf-measures/type-check`, don't you?

Yes. Exactly!

**Comment by m-shaka** [comment]
@yusukebe moved the directory and made some refactoring!

File: perf-measures/type-check/.gitignore
Comment (yusukebe) [inline perf-measures/type-check/.gitignore:1]: How about adding a `result.txt`?

Comment (m-shaka) [reply perf-measures/type-check/.gitignore:1]: Done!

**Review by yusukebe — APPROVED** [review]
LGTM!

**Comment by yusukebe** [comment]
@m-shaka 

Thanks! Let's go with this and keep improving it.

**Pull request merged**